llvm/lib/Bitcode/Writer/ValueEnumerator.cpp

```diff
@@ -261,33 +269,39 @@ static UseListOrderStack predictUseListOrder(const Module &M) {
   // constants in the last Function they're used in.  Module-level constants
   // have already been visited above.
   for (const Function &F : llvm::reverse(M)) {
+    auto PredictValueOrderFromMetadata = [&](Metadata *MD) {
+      if (const auto *VAM = dyn_cast<ValueAsMetadata>(MD)) {
+        predictValueUseListOrder(VAM->getValue(), &F, OM, Stack);
+      } else if (const auto *AL = dyn_cast<DIArgList>(MD)) {
+        for (const auto *VAM : AL->getArgs())
+          predictValueUseListOrder(VAM->getValue(), &F, OM, Stack);
+      }
+    };
```

Comment:
This is a good question. I added it because we do it for debug intrinsics already (see below, peering through ValueAsMetadatas). I think you might be right, that it's not needed. But I'm not 100% certain. I'm going to push the rest of the changes for now and dig deeper into this.

Reply:
Is it possible that this is flushing out the order of the use list /inside/ the MAV/VAM pairing? I think it's technically possible for that to become visible, if you `findDbgUsers` on a value then move all of it's users to a position, you enumerate them in the use-list order. It's not clear that that's what this code preserves though (perhaps nothing preserves it).

Reply:
Nah I don't think so. The MAV/VAM is unwrapped, and predictValueUseListOrder only maps `uses()` of the Value, so the MAV/VAM "use" isn't considered at all.

Reply:
Keeping things in-sync makes sense to me; avoids any unexpected differences.